overlord/servicestate/servicemgr.go

```diff
@@ -42,8 +55,241 @@ func Manager(st *state.State, runner *state.TaskRunner) *ServiceManager {
+			}
+		}
+	}
+
+	// Second loop actually restarts the services per-snap by sorting them and
+	// removing disabled services. Note that we could have disabled services
+	// here because a service could have been running, but disabled when snapd
+	// was refreshed, hence it got killed, but we don't want to restart it,
+	// since it is disabled, and so that disabled running service is just SOL.
+	for sn, apps := range candidateAppsToRestartBySnap {
+		// TODO: should we try to start as many services as possible here before
+		// giving up given the severity of the bug?
+		disabledSvcs, err := wrappers.QueryDisabledServices(sn, progress.Null)
+		if err != nil {
+			return err
+		}
+
+		startupOrdered, err := snap.SortServices(apps)
+		if err != nil {
+			return err
+		}
+
+		// TODO: what to do about timings here?
+		nullPerfTimings := &timings.Timings{}
+		if err := wrappers.StartServices(startupOrdered, disabledSvcs, nil, progress.Null, nullPerfTimings); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *ServiceManager) ensureSnapServicesUpdated() (err error) {
+	m.state.Lock()
+	defer m.state.Unlock()
+	if m.ensuredSnapSvcs {
+		return nil
+	}
+
+	// only run after we are seeded
+	var seeded bool
+	err = m.state.Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	if !seeded {
+		return nil
+	}
+
+	// we are seeded, now we need to find all snap services and re-generate
+	// services as necessary
+
+	// ensure all snap services are updated
+	allStates, err := snapstate.All(m.state)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	// if we have no snaps we can exit early
+	if len(allStates) == 0 {
+		m.ensuredSnapSvcs = true
+		return nil
+	}
+
+	snapsMap := map[*snap.Info]*wrappers.SnapServiceOptions{}
+
+	for _, snapSt := range allStates {
+		info, err := snapSt.CurrentInfo()
+		if err != nil {
+			return err
+		}
+
+		// don't use EnsureSnapServices with the snapd snap
+		if info.Type() == snap.TypeSnapd {
+			continue
+		}
+
+		snapSvcOpts, err := SnapServiceOptions(m.state, info.InstanceName())
+		if err != nil {
+			return err
+		}
+		snapsMap[info] = snapSvcOpts
+	}
+
+	// setup ensure options
+	ensureOpts := &wrappers.EnsureSnapServicesOptions{
+		Preseeding: snapdenv.Preseeding(),
+	}
+
+	// set RequireMountedSnapdSnap if we are on UC18+ only
+	deviceCtx, err := snapstate.DeviceCtx(m.state, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	if !deviceCtx.Classic() && deviceCtx.Model().Base() != "" {
+		ensureOpts.RequireMountedSnapdSnap = true
+	}
+
+	// TODO: should we use an actual interacter here ?
+
+	// TODO: this should take a callback so that we can gate the logic about
+	// restarting services below on actually observing a Requires= to Wants=
+	// transition which is what we need to do to fix LP #1924805
+	modified, err := wrappers.EnsureSnapServices(snapsMap, ensureOpts, progress.Null)
```

Comment:
we have a problem with modified that I noticed in my PR about callbacks,  it will contains the same app multiple times if something like both the services and sockets/timers are modified at the same time

Reply:
Hm timers and sockets should not change. Do you know why the content is different?

Reply:
This was discussed in MM and it's all clear now.